File: android/src/main/kotlin/com/hendrick/plugin_workshop/PluginWorkshopPlugin.kt
package com.hendrick.plugin_workshop

import ArithmeticFlutterApi
import ArithmeticHostApi
import ArithmeticOperation
import PluginPigeonError
import io.flutter.embedding.engine.plugins.FlutterPlugin
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.launch
import java.util.Timer
import kotlin.concurrent.timer

/** PluginWorkshopPlugin */
class PluginWorkshopPlugin : FlutterPlugin, ArithmeticHostApi {

    private var apiTimer: Timer? = null
    private var flutterApi: ArithmeticFlutterApi? = null
    private var result: Long = 0

    private val mainScope = MainScope()

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        ArithmeticHostApi.setUp(flutterPluginBinding.binaryMessenger, this)
        flutterApi = ArithmeticFlutterApi(flutterPluginBinding.binaryMessenger)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        ArithmeticHostApi.setUp(binding.binaryMessenger, null)
        flutterApi = null
    }


    override fun performArithmeticOperation(
        input1: Double, input2: Double, operation: ArithmeticOperation
    ): Double {
        return when (operation) {
            ArithmeticOperation.ADD -> input1 + input2
            ArithmeticOperation.SUBTRACT -> input1 - input2
            ArithmeticOperation.MULTIPLY -> input1 * input2
            ArithmeticOperation.DIVIDE -> {
                if (input2 == 0.0) {
                    throw PluginPigeonError("DIVISION_BY_ZERO", "Cannot divide by zero")
                }

                return input1 / input2
            }
        }
    }

    override fun startTimer() {
        apiTimer?.cancel()
        apiTimer = timer(period = 2000L) {
            mainScope.launch {
                flutterApi?.onReceiveTimerResult(result) {
                    if (it.isSuccess) {
                        result += 1
                    }
                }
            }
        }
    }

    override fun stopTimer() {
        apiTimer?.cancel()
        apiTimer = null
        result = 0

        flutterApi?.onReceiveTimerResult(result) {}
    }
}

File: ios/Classes/PluginWorkshopPlugin.swift
import Flutter
import UIKit

public class PluginWorkshopPlugin: NSObject, FlutterPlugin, ArithmeticHostApi {
    var apiTimer: Timer? = nil
    var flutterApi: ArithmeticFlutterApi? = nil
    var result: Int64 = 0
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let messenger = registrar.messenger()
        let api = PluginWorkshopPlugin.init()
        ArithmeticHostApiSetup.setUp(binaryMessenger: messenger, api: api)
        api.flutterApi = ArithmeticFlutterApi(binaryMessenger: messenger)
    }
    
    public func detachFromEngine(for registrar: any FlutterPluginRegistrar) {
        ArithmeticHostApiSetup.setUp(binaryMessenger: registrar.messenger(), api: nil)
        self.flutterApi = nil
    }
    
    func performArithmeticOperation(input1: Double, input2: Double, operation: ArithmeticOperation) throws -> Double {
        switch(operation) {
        case .add:
            return input1 + input2
        case .subtract:
            return input1 - input2
        case .multiply:
            return input1 * input2
        case .divide:
            if (input2 == 0.0) {
                throw PluginPigeonError(code: "DIVISION_BY_ZERO", message: "Cannot divide by zero", details: nil)
            }
            
            return input1 / input2
        }
    }
    
    func startTimer() throws {
        apiTimer?.invalidate()
        apiTimer = Timer.scheduledTimer(withTimeInterval: 2, repeats: true, block: { _ in
            self.flutterApi?.onReceiveTimerResult(result: self.result, completion: { result in
                switch(result) {
                case .success():
                    self.result += 1
                case .failure(let errorResult):
                    print("an error occurred: \(errorResult)")
                }
            })
        })
    }
    
    func stopTimer() throws {
        apiTimer?.invalidate()
        apiTimer = nil
        result = 0
        
        self.flutterApi?.onReceiveTimerResult(result: 0, completion: { _  in})
    }
}
